# Praca domowa 2: pomiar pamięci i czasu obliczeń w notebooku Jupyter





In [17]:
import sys
import time

lista_1 = list(range(50))
lista_2 = list(range(500))
lista_3 = list(range(5000))
lista_4 = list(range(5000000))


print(f"rozmiar 0-49:" , sys.getsizeof(lista_1), "bajtow")
print(f"rozmiar 0-499:" , sys.getsizeof(lista_2), "bajtow")
print(f"rozmiar 0-4999:" , sys.getsizeof(lista_3), "bajtow")
print(f"rozmiar 0-499999:" , sys.getsizeof(lista_4), "bajtow")

rozmiar 0-49: 456 bajtow
rozmiar 0-499: 4056 bajtow
rozmiar 0-4999: 40056 bajtow
rozmiar 0-499999: 40000056 bajtow


In [18]:
#sposób2
start = time.time()

srednia = sum(lista_4) / len(lista_4)

koniec = time.time()
wynik = koniec - start

print(f"Obliczona srednia: {srednia}")
print(f"Czas trwania obliczeń dla 4,99999 milionów liczb: {wynik:.5f} sekund")

Obliczona srednia: 2499999.5
Czas trwania obliczeń dla 4,99999 milionów liczb: 0.17804 sekund


## Wyniki pomiarów pamięci i czasu dla list

Wnioski:

* **Pamięć:** Zużycie pamięci rośnie równomiernie wraz ze wzrostem liczby elementów. Lista 50 elementów zajmuje 456 bajtów, a lista 5 000 000 elementów zajmuje już aż 40 000 056 bajtów (ok. 40 megabajtów). Każdy dodatkowy element w liście zwiększa jej rozmiar w pamięci o stałą liczbę bajtów (potrzebną na przechowanie wskaźnika).
* **Czas wykonania:** Dla mniejszych list (50, 500, 5000) czas obliczania średniej za pomocą `sum() / len()` jest niezauważalny dla człowieka (rząd milionowych części sekundy).
* **Zależność od liczby elementów:** Przy skoku do 5 milionów elementów czas wyraźnie rośnie - komputer potrzebował około 0.136 sekundy. Tradycyjna metoda wymaga fizycznego przejrzenia i dodania do siebie każdej liczby po kolei. Im większa lista, tym więcej operacji musi wykonać procesor.

In [19]:
#wszystkie przedziały
pakiety_list = [lista_1, lista_2, lista_3, lista_4]

for i, lst in enumerate(pakiety_list):
    start = time.time()
    srednia = sum(lst) / len(lst)
    koniec = time.time()
    czas = koniec - start
    print(f"kista {len(lst)} elementów, Średnia: {srednia},  Czas: {czas:.10f} s")

kista 50 elementów, Średnia: 24.5,  Czas: 0.0265841484 s
kista 500 elementów, Średnia: 249.5,  Czas: 0.0000228882 s
kista 5000 elementów, Średnia: 2499.5,  Czas: 0.0001380444 s
kista 5000000 elementów, Średnia: 2499999.5,  Czas: 0.2268941402 s


## Przykład na szybsze obliczenie średniej



* Dla $n = 5$ (liczby: 0, 1, 2, 3, 4) -> Suma = 10. Średnia = 10 / 5 = **2.0**
* Dla $n = 7$ (liczby: 0, 1, 2, 3, 4, 5, 6) -> Suma = 21. Średnia = 21 / 7 = **3.0**
* Dla $n = 10$ (liczby: 0 do 9) -> Suma = 45. Średnia = 45 / 10 = **4.5**

**Wniosek:** Średnia z takiego zestawu liczb to zawsze dokładny środek tego przedziału. Matematycznie, średnia z ciągu kolejnych liczb to po prostu suma pierwszej i ostatniej liczby podzielona przez 2. Ponieważ zaczynamy od 0, a kończymy na n-1, wzór na średnią wygląda następująco:

$$\text{Średnia} = \frac{0 + (n - 1)}{2} = \frac{n - 1}{2}$$

Dzięki temu wzorowi możemy obliczyć średnią w jednym kroku.

In [20]:
#sposób3
n = 5000000

start_2 = time.time()

srednia_szybka = (n - 1) / 2

koniec_2 = time.time()
czas_szybki = koniec_2- start_2

print(f"Obliczona średnia wzorem: {srednia_szybka}")
print(f"Czas trwania nowego sposobu: {czas_szybki:.10f} sekund")

Obliczona średnia wzorem: 2499999.5
Czas trwania nowego sposobu: 0.0000348091 sekund


In [21]:
#wszystkie przedziały
przedzialy = [50, 500, 5000, 5000000]

for n in przedzialy:
    start = time.time()
    srednia = (n - 1) / 2
    koniec = time.time()
    czas = koniec - start
    print(f"Przedział do {n-1} Średnia {srednia}  Czas: {czas:.8f} s")

Przedział do 49 Średnia 24.5  Czas: 0.00000167 s
Przedział do 499 Średnia 249.5  Czas: 0.00000215 s
Przedział do 4999 Średnia 2499.5  Czas: 0.00000000 s
Przedział do 4999999 Średnia 2499999.5  Czas: 0.00000095 s


## Wnioski końcowe - Porównanie metod

Po przeanalizowaniu obu podejść, możemy sformułować końcowe podsumowanie:

1.  **Pamięciożerność:** * **Metoda z listą:** Jest bardzo pamięciożerna. Wymaga zapisania wszystkich elementów w pamięci RAM  5 milionów liczb to ok. 40 MB.
    * **Metoda matematyczna:** Nie potrzebuje listy ani pamięci RAM na kolekcję danych – przechowuje tylko jedną liczbę n, więc zużycie pamięci wynosi praktycznie zero.
2.  **Szybkość (Czas obliczeń):** * **Metoda z listą:** Zwalnia wraz ze wzrostem danych. Dla 5 milionów liczb trwało to ponad 0.13 sekundy.
    * **Metoda matematyczna:** Działa natychmiastowo. Z pomiarów wynika, że czas dla 5 milionów liczb spadł do wartości bliskich zeru (0.00000095 s). Czas działania jest stały i niezależny od wielkości liczby n.
3.  **Kiedy tworzenie listy ma sens?**
    Tworzenie listy i ręczne liczenie sumy oraz średniej ma sens wtedy, gdy dane są nieregularne, losowe i nie pochodzą z żadnego ciągu (np. dane pogodowe).
4.  **Kiedy lepiej wykorzystać własność matematyczną?**
    Zawsze, gdy dane mają znaną strukturę (np. są kolejnymi liczbami, ciągami arytmetycznymi lub geometrycznymi). Wykorzystanie wzoru pozwala zaoszczędzić gigantyczne ilości pamięci i czasu procesora.